patches/server/0728-Add-more-Campfire-API.patch

```diff
@@ -0,0 +1,105 @@
+                 int j = campfire.cookingProgress[i]++;
++                } // Paper
+ 
+                 if (campfire.cookingProgress[i] >= campfire.cookingTime[i]) {
+                     SimpleContainer inventorysubcontainer = new SimpleContainer(new ItemStack[]{itemstack});
+@@ -156,6 +160,12 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
+             System.arraycopy(aint, 0, this.cookingTime, 0, Math.min(this.cookingTime.length, aint.length));
+         }
+ 
++        // Paper start
++        if (nbt.contains("Paper.StopCooking", 7)) { // 7: Byte Array
++            byte[] abyte = nbt.getByteArray("Paper.StopCooking");
++            System.arraycopy(abyte, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, abyte.length));
++        }
++        // Paper end
+     }
+ 
+     @Override
+@@ -163,6 +173,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
+         this.saveMetadataAndItems(nbt);
+         nbt.putIntArray("CookingTimes", this.cookingProgress);
+         nbt.putIntArray("CookingTotalTimes", this.cookingTime);
++        nbt.putByteArray("Paper.StopCooking", this.stopCooking); // Paper
+         return nbt;
+     }
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java b/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
+index 5ac24016f13f0739425c95861f6337a930d3d08e..01a4a9ca54886d53c173aa8f7a28f3d01d8d2f52 100644
+--- a/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
++++ b/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
+@@ -52,4 +52,44 @@ public class CraftCampfire extends CraftBlockEntityState<CampfireBlockEntity> im
+     public void setCookTimeTotal(int index, int cookTimeTotal) {
+         getSnapshot().cookingTime[index] = cookTimeTotal;
+     }
++
++    // Paper start
++    @Override
++    public void disableCooking() {
++        for (int i = 0; i < getSnapshot().stopCooking.length; ++i)
++            this.disableCooking(i);
++    }
++
++    @Override
++    public void enableCooking() {
++        for (int i = 0; i < getSnapshot().stopCooking.length; ++i)
++            this.enableCooking(i);
++    }
++
++    @Override
++    public boolean disableCooking(int index) {
++        boolean previous = this.isDisableCooking(index);
++        getSnapshot().stopCooking[index] = (byte) 1;
++        return previous;
++    }
++
++    @Override
++    public boolean enableCooking(int index) {
++        boolean previous = this.isDisableCooking(index);
++        getSnapshot().stopCooking[index] = (byte) 0;
++        return previous;
++    }
++
++    @Override
++    public boolean isDisableCooking() {
++        for (int i = 0; i < getSnapshot().stopCooking.length; ++i)
++            if (!this.isDisableCooking(i)) return false;
++        return true;
++    }
++
++    @Override
++    public boolean isDisableCooking(int index) {
++        return getSnapshot().stopCooking[index] == (byte) 1;
```

Comment:
Have a Preconditions check above this line to make sure the `index` parameter isn't going to through an OutOfBounds exception. Like if I pass in -3.